client/service/src/lib.rs

```diff
@@ -380,57 +380,35 @@ where
 		}
 	}
 
-	let (max_request_size, ws_max_response_size, http_max_response_size) =
-		legacy_cli_parsing(config);
-
 	let random_port = |mut addr: SocketAddr| {
 		addr.set_port(0);
 		addr
 	};
 
-	let ws_addr = config
-		.rpc_ws
+	let addr = config
+		.rpc_addr
 		.unwrap_or_else(|| "127.0.0.1:9944".parse().expect("valid sockaddr; qed"));
-	let ws_addr2 = random_port(ws_addr);
-
-	let http_addr = config
-		.rpc_http
-		.unwrap_or_else(|| "127.0.0.1:9933".parse().expect("valid sockaddr; qed"));
-	let http_addr2 = random_port(http_addr);
+	let addr2 = random_port(addr);
 
 	let metrics = sc_rpc_server::RpcMetrics::new(config.prometheus_registry())?;
 
-	let server_config = sc_rpc_server::WsConfig {
-		max_connections: config.rpc_ws_max_connections,
-		max_payload_in_mb: max_request_size,
-		max_payload_out_mb: ws_max_response_size,
+	let server_config = sc_rpc_server::Config {
+		addrs: [addr, addr2],
+		max_connections: config.rpc_max_connections,
+		max_payload_in_mb: config.rpc_max_request_size,
+		max_payload_out_mb: config.rpc_max_response_size,
 		max_subs_per_conn: config.rpc_max_subs_per_conn,
+		rpc_api: gen_rpc_module(deny_unsafe(addr, &config.rpc_methods))?,
+		metrics,
+		id_provider: rpc_id_provider,
+		cors: config.rpc_cors.as_ref(),
+		tokio_handle: config.tokio_handle.clone(),
 	};
 
-	let http_fut = sc_rpc_server::start_http(
-		[http_addr, http_addr2],
-		config.rpc_cors.as_ref(),
-		max_request_size,
-		http_max_response_size,
-		metrics.clone(),
-		gen_rpc_module(deny_unsafe(http_addr, &config.rpc_methods))?,
-		config.tokio_handle.clone(),
-	);
-
-	let ws_fut = sc_rpc_server::start(
-		[ws_addr, ws_addr2],
-		config.rpc_cors.as_ref(),
-		server_config.clone(),
-		metrics.clone(),
-		gen_rpc_module(deny_unsafe(ws_addr, &config.rpc_methods))?,
-		config.tokio_handle.clone(),
-		rpc_id_provider,
-	);
-
-	match tokio::task::block_in_place(|| {
-		config.tokio_handle.block_on(futures::future::try_join(http_fut, ws_fut))
-	}) {
-		Ok((http, ws)) => Ok(Box::new((waiting::Server(Some(http)), waiting::Server(Some(ws))))),
+	let server_fut = sc_rpc_server::start_server(server_config);
+
+	match tokio::task::block_in_place(|| config.tokio_handle.block_on(server_fut)) {
+		Ok(server) => Ok(Box::new(waiting::Server(Some(server)))),
 		Err(e) => Err(Error::Application(e)),
 	}
```

Comment:
```suggestion
	match config.tokio_handle.block_on(server_fut) {
		Ok(server) => Ok(Box::new(waiting::Server(Some(server)))),
		Err(e) => Err(Error::Application(e)),
	}
```

We don't really need `block_in_place` here? Maybe we should also make the entire function async

Reply:
I like the idea of making the entire function async but IIRC I didn't want to starve other spawned tasks but given that this is on startup it is probably fine.

I'll have look

Reply:
Yeah we would just move the block_on call. I think we should block to ensure that we bring down the node if the RPC server failed to start.

So, TLDR: Maybe the best is to only apply my suggested changes.

Reply:
added a comment, we need it because some tests calls this function and already blocks on it in.